docs/tutorials/data_provider.md

```diff
@@ -102,13 +102,11 @@ impl<'a> Borrow<CacheKey<'a>> for lru::KeyRef<CacheKeyWrap> {
     }
 }
 
-impl<M, P> DataProvider<M> for LruDataCache<P>
+impl<M: KeyedDataMarker, P: DataProvider<M>> DataProvider<M> for LruDataCache<P>
```

Comment:
Issue: Throughout this PR you are applying a style change of moving bounds from `where` to `impl<>`, in addition to changing the `'static`. There are two problems with that: (1) the PR is larger in scope than originally planned, and (2) I am not in alignment that this is a good change; personally I prefer to have all bounds in `where` or all in `impl` but not a mix. We should probably write this down in the style guide, but this PR is not the place to discuss that. Please make this PR smaller by only changing `'static`

Reply:
The heuristic I'm usually using is that if the bound is a simple `Generic: Trait`, it would go inline. If it's multiple bounds or bounds on associated types, it would not. This way if you have bounds like:

```rust
impl<M> DataProvider<M> for AnyPayloadProvider
where
    M: KeyedDataMarker,
    for<'a> YokeTraitHack<<M::Yokeable as Yokeable<'a>>::Output>: Clone,
    M::Yokeable: ZeroFrom<'static, M::Yokeable>,
    M::Yokeable: icu_provider::MaybeSendSync,
```

you wouldn't have to go into the verbose `where` clause to figure out what `M` is.

Removing the `'static` makes these bounds simple, so I would inline them.

Reply:
FWIW the heuristic @robertbastian mentions is quite commonly found in Rust, specifically:

 - impl bounds tell you what a type parameter _is_ (e.g. "i primarily care about this as a serializable type" or "this is a marker type")
 - where bounds tell you about other constraints

But I agree that we shouldn't do this in this PR, and I'm iffy on the exact rule here.